File: app/src/main/java/com/jonnyhsia/composer/widget/DividerVertical.java
package com.jonnyhsia.composer.widget;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.support.v7.widget.RecyclerView;
import android.view.View;

/**
 * RecyclerView 纵向布局简单分割线
 */
public class DividerVertical extends RecyclerView.ItemDecoration {

    private Paint mPaint;

    /**
     * 是否绘制列表顶部的分割线
     */
    private boolean mDrawListTopDivider;

    /**
     * 是否绘制列表底部的分割线
     */
    private boolean mDrawListBottomDivider;
    private int dividerHeight;

    public DividerVertical(boolean drawTop, boolean drawBottom, int dividerHeight) {
        this(drawTop, drawBottom, dividerHeight, Color.TRANSPARENT);
    }

    public DividerVertical(boolean drawTop, boolean drawBottom, int dividerHeight, int color) {
        this.mDrawListTopDivider = drawTop;
        this.mDrawListBottomDivider = drawBottom;
        this.dividerHeight = dividerHeight;
        mPaint = new Paint();
        mPaint.setColor(color);
        mPaint.setStyle(Paint.Style.FILL);
    }

    @Override
    public void getItemOffsets(Rect outRect, View view, RecyclerView parent, RecyclerView.State state) {
        int position = parent.getChildLayoutPosition(view);
        if (mDrawListTopDivider && mDrawListBottomDivider) {
            if (position == 0) {
                outRect.set(0, dividerHeight, 0, dividerHeight);
            } else {
                outRect.set(0, 0, 0, dividerHeight);
            }

            return;
        }
        if (mDrawListTopDivider) {
            outRect.set(0, dividerHeight, 0, 0);
            return;
        }
        if (mDrawListBottomDivider) {
            outRect.set(0, 0, 0, dividerHeight);
            return;
        }
        if (position == 0)
            outRect.set(0, 0, 0, 0);
        else
            outRect.set(0, dividerHeight, 0, 0);
    }

    @Override
    public void onDraw(Canvas c, RecyclerView parent, RecyclerView.State state) {
        super.onDrawOver(c, parent, state);
        final int left = parent.getPaddingLeft();
        final int right = parent.getWidth() - parent.getPaddingRight();

        final int count = parent.getChildCount();
        for (int i = 0; i < count - 1; i++) {
            final View child = parent.getChildAt(i);
            RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) child.getLayoutParams();
            final int top = child.getBottom() + params.bottomMargin
                    + Math.round(child.getTranslationY());
            final int bottom = top + dividerHeight;
            c.drawRect(left, top, right, bottom, mPaint);
        }

        //第一条分割线
        if (mDrawListTopDivider)
            c.drawRect(left
                    , parent.getPaddingTop()
                    , right
                    , parent.getPaddingTop() + dividerHeight
                    , mPaint);
        if (mDrawListBottomDivider)
            c.drawRect(left
                    , parent.getBottom() - parent.getPaddingBottom() - dividerHeight
                    , right
                    , parent.getBottom() - parent.getPaddingBottom()
                    , mPaint);
    }
}

File: app/src/main/java/com/jonnyhsia/composer/widget/DividerVertical.kt
package com.jonnyhsia.composer.widget

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.support.v7.widget.RecyclerView
import android.view.View

/**
 * RecyclerView 纵向布局简单分割线
 */
class DividerVertical @JvmOverloads constructor(
        private val dividerHeight: Int,
        dividerColor: Int = Color.TRANSPARENT,
        private val drawTop: Boolean = false,
        private val drawBottom: Boolean = false
) : RecyclerView.ItemDecoration() {

    private val paint: Paint = Paint().apply {
        color = dividerColor
        style = Paint.Style.FILL
    }

    override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State?) {
        val position = parent.getChildLayoutPosition(view)
        if (drawTop && drawBottom) {
            if (position == 0) {
                outRect.set(0, dividerHeight, 0, dividerHeight)
            } else {
                outRect.set(0, 0, 0, dividerHeight)
            }

            return
        }
        if (drawTop) {
            outRect.set(0, dividerHeight, 0, 0)
            return
        }
        if (drawBottom) {
            outRect.set(0, 0, 0, dividerHeight)
            return
        }
        if (position == 0)
            outRect.set(0, 0, 0, 0)
        else
            outRect.set(0, dividerHeight, 0, 0)
    }

    override fun onDraw(c: Canvas, parent: RecyclerView, state: RecyclerView.State?) {
        super.onDrawOver(c, parent, state)
        val left = parent.paddingLeft
        val right = parent.width - parent.paddingRight

        val count = parent.childCount
        for (i in 0 until count - 1) {
            val child = parent.getChildAt(i)
            val params = child.layoutParams as RecyclerView.LayoutParams
            val top = (child.bottom + params.bottomMargin
                    + Math.round(child.translationY))
            val bottom = top + dividerHeight
            c.drawRect(left.toFloat(), top.toFloat(), right.toFloat(), bottom.toFloat(), paint)
        }

        //第一条分割线
        if (drawTop)
            c.drawRect(left.toFloat(), parent.paddingTop.toFloat(), right.toFloat(), (parent.paddingTop + dividerHeight).toFloat(), paint)
        if (drawBottom)
            c.drawRect(left.toFloat(), (parent.bottom - parent.paddingBottom - dividerHeight).toFloat(), right.toFloat(), (parent.bottom - parent.paddingBottom).toFloat(), paint)
    }
}
